python/felis/datamodel.py

```diff
@@ -0,0 +1,354 @@
+    """The IVOA TAP table index of the table."""
+
+    mysql_engine: str | None = Field(None, alias="mysql:engine")
+    """The mysql engine to use for the table.
+
+    For now this is a freeform string but it could be constrained to a list of
+    known engines in the future.
+    """
+
+    mysql_charset: str | None = Field(None, alias="mysql:charset")
+    """The mysql charset to use for the table.
+
+    For now this is a freeform string but it could be constrained to a list of
+    known charsets in the future.
+    """
+
+    @model_validator(mode="before")
+    @classmethod
+    def create_constraints(cls, values: dict[str, Any]) -> dict[str, Any]:
+        """Create constraints from the `constraints` field."""
+        if "constraints" in values:
+            new_constraints: list[Constraint] = []
+            for item in values["constraints"]:
+                if item["@type"] == "ForeignKey":
+                    new_constraints.append(ForeignKeyConstraint(**item))
+                elif item["@type"] == "Unique":
+                    new_constraints.append(UniqueConstraint(**item))
+                elif item["@type"] == "Check":
+                    new_constraints.append(CheckConstraint(**item))
+                else:
+                    raise ValueError(f"Unknown constraint type: {item['@type']}")
+            values["constraints"] = new_constraints
+        return values
+
+    @field_validator("columns", mode="after")
+    @classmethod
+    def check_unique_column_names(cls, columns: list[Column]) -> list[Column]:
+        """Check that column names are unique."""
+        if len(columns) != len(set(column.name for column in columns)):
+            raise ValueError("Column names must be unique")
+        return columns
+
+
+class SchemaVersion(BaseModel):
+    """The version of the schema."""
+
+    current: str
+    """The current version of the schema."""
+
+    compatible: list[str] | None = None
+    """The compatible versions of the schema."""
+
+    read_compatible: list[str] | None = None
+    """The read compatible versions of the schema."""
+
+
+def _extract_ids(obj: Any, id_map: dict[str, Any], processed: set[int]) -> None:
+    """Recursively extract ID values from a `BaseModel` object
+    and map them to their corresponding object.
+    """
+    obj_id = id(obj)
+    if obj_id in processed:
+        return
+    processed.add(obj_id)
+
+    if isinstance(obj, BaseModel):
+        if hasattr(obj, "id"):
+            id_map[getattr(obj, "id")] = obj
+
+        # Iterate over the fields of the BaseModel object
+        for attr, value in obj.__dict__.items():
+            if isinstance(value, BaseModel | list | dict):
+                _extract_ids(value, id_map, processed)
+
+    elif isinstance(obj, list):
+        for item in obj:
+            _extract_ids(item, id_map, processed)
+
+    elif isinstance(obj, dict):
+        for value in obj.values():
+            _extract_ids(value, id_map, processed)
```

Comment:
Do you need to add a final `else` block here for unexpected content? mypy isn't going to help you because you declared `obj` as `Any` rather than `BaseModel | list[BaseModel] | dict[str, BaseModel]`.

Reply:
If there were unexpected content, the validation should fail, which means this method should never be run, as it is designated as "after." 

But I like the typing suggestion and will implement it.